Data/models/Bookmark.swift

```diff
@@ -96,59 +96,62 @@ public final class Bookmark: NSManagedObject, WebsitePresentable, Syncable, CRUD
     // Syncable
     public func update(syncRecord record: SyncRecord?) {
         guard let bookmark = record as? SyncBookmark, let site = bookmark.site else { return }
-        title = site.title
-        update(customTitle: site.customTitle, url: site.location, newSyncOrder: bookmark.syncOrder)
-        lastVisited = Date(timeIntervalSince1970: (Double(site.lastAccessedTime ?? 0) / 1000.0))
-        syncParentUUID = bookmark.parentFolderObjectId
-        // No auto-save, must be handled by caller if desired
+        
+        DataController.performBackgroundTask { context in
+            
+            guard let bookmarkToUpdate = context.object(with: self.objectID) as? Bookmark else { return }
+            
+            bookmarkToUpdate.title = site.title
+            bookmarkToUpdate.update(customTitle: site.customTitle, url: site.location,
+                                    newSyncOrder: bookmark.syncOrder, context: context)
+            bookmarkToUpdate.lastVisited = Date(timeIntervalSince1970: (Double(site.lastAccessedTime ?? 0) / 1000.0))
+            bookmarkToUpdate.syncParentUUID = bookmark.parentFolderObjectId
+        }
     }
     
     public func update(customTitle: String?, url: String?, newSyncOrder: String? = nil, save: Bool = false,
-                       sendToSync: Bool = false) {
-        var contextToUpdate: NSManagedObjectContext?
-        
-        // Syncable.update() doesn't save to CD at the moment, we need to use managedObjectContext here.
-        if save == false {
-            contextToUpdate = managedObjectContext
-            // Updated object usually uses view context, all database writes should happen on a background thread
-            // so we need to fetch the object using background context.
-        } else if managedObjectContext?.concurrencyType != .privateQueueConcurrencyType {
-            contextToUpdate = DataController.newBackgroundContext()
-        }
+                       sendToSync: Bool = false, context: NSManagedObjectContext? = nil) {
         
-        guard let bookmarkToUpdate = (try? contextToUpdate?.existingObject(with: objectID)) as? Bookmark,
-            let context = contextToUpdate else {
-                return
+        if let context = context {
```

Comment:
This is a pattern I use in many places in code. It's supposed to prevent launching multiple background tasks for a given code. 

When context is provided we never have to worry about any threading because the  code is already wrapped in a safe thread.
When context is not provided we can then safely launch a new background task.

This solution is not perfect because you need to be aware of passing correct context all the time, as not providing any context results in launching a new thread.

Reply:
I think we should put in an assert
`assert(!isMainThread, "No saving updating on main thread")`
And use this liberally, assuming there are no more known instances of updating on main thread.

If this is not on the mainthread it can then be reduced more simply to:
```
let performingContext = context ?? DataController.newBackgroundContext
DataController.perform(with: performingContext) { // See convo on `performBackgroundTask` to understand this
     guard let bookmarkToUpdate = performingContext.object(with: self.objectID) as? Bookmark else { return }
     bookmarkToUpdate.updateImpl(customTitle: customTitle, url: url, newSyncOrder: newSyncOrder, save: save,
                           sendToSync: sendToSync, context: backgroundContext)
}
```

AFAIK, this is true for most of the other instances around too.